File: python/algo/202511/algo0398.py
from collections import defaultdict
import random
from typing import List

# 数学算法 - 随机算法 - 6.1 随机数
class Solution:
    # 自行解答 随机函数
    def __init__(self, nums: List[int]):
        self.n = len(nums)
        self.t_inxs = defaultdict(list)
        for i, v in enumerate(nums): self.t_inxs[v].append(i)
        random.seed(100)
            

    def pick(self, target: int) -> int:
        inx = random.randint(0, self.n) % len(self.t_inxs[target])
        return self.t_inxs[target][inx]
    

File: cplus/leetcode/202511/algo0398.cpp
#include <bits/stdc++.h>
using namespace std;

// 数学算法 - 随机算法 - 6.1 随机数
class Solution {
private:
    unordered_map<int, vector<int>> vals;
public:
    Solution(vector<int>& nums) {
        for (int i=0; i<nums.size(); ++i) vals[nums[i]].emplace_back(i);
    }
    
    int pick(int target) {
        return vals[target][rand() % vals[target].size()];
    }
};